Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -397,36 +397,100 @@ private class AttributeRemover: SyntaxRewriter {
+        var leadingTrivia = attribute.leadingTrivia
+        if let lastNewline = leadingTrivia.pieces.lastIndex(where: \.isNewline),
+          attribute.trailingTrivia.isEmpty,
+          let nextToken = attribute.nextToken(viewMode: .sourceAccurate),
+          !nextToken.leadingTrivia.isEmpty
         {
           // If the attribute is on its own line based on the following conditions,
           // remove the newline from it so we don’t end up with an empty line
-          //  - Trailing trivia ends with a newline followed by arbitrary number of spaces or tabs
-          //  - There is no trailing trivia and the next token starts on a new line
+          //  - Leading trivia ends with a newline followed by arbitrary number of spaces or tabs
+          //  - There is no trailing trivia and the next token has no leading trivia
           leadingTrivia = Trivia(pieces: leadingTrivia.pieces[..<lastNewline])
         }
         // Drop any spaces or tabs from the trailing trivia because there’s no
         // more attribute they need to separate.
-        let trailingTrivia = Trivia(pieces: attribute.trailingTrivia.drop(while: { $0.isSpaceOrTab }))
+        let trailingTrivia = attribute.trailingTrivia.trimmingPrefix(while: \.isSpaceOrTab)
         triviaToAttachToNextToken += leadingTrivia + trailingTrivia
+
+        // If the attribute is not separated by the previous attribute using trivia, e.g.
+        // `@First@Second var x: Int` (yes, that's valid Swift), and removing  the `@Second`
+        // attribute then dropping all trivia of this attribute would join `@First` and `var`
+        // without any trivia in between, which is invalid. In this case the trailing trivia of the
+        // attribute is significant and we need to keep it.
+        if triviaToAttachToNextToken.isEmpty,
+          let previousToken = attribute.previousToken(viewMode: .sourceAccurate),
+          previousToken.trailingTrivia.isEmpty
+        {
+          triviaToAttachToNextToken = attribute.trailingTrivia
+        }
+      } else {
+        filteredAttributes.append(.attribute(prependAndClearAccumulatedTrivia(to: attribute)))
+      }
+    }
+
+    // Ensure that any horizontal whitespace trailing the attributes list is trimmed if the next
+    // token starts a new line.
+    if let nextToken = node.nextToken(viewMode: .sourceAccurate),
+      nextToken.leadingTrivia.startsWithNewline()
+    {
+      if !triviaToAttachToNextToken.isEmpty {
+        triviaToAttachToNextToken = triviaToAttachToNextToken.trimmingSuffix(while: \.isSpaceOrTab)
       } else {
-        filteredAttributes.append(.attribute(attribute))
+        if let last = filteredAttributes.last {
+          filteredAttributes =
+            filteredAttributes.dropLast()
+            + [
+              last.with(
+                \.trailingTrivia,
+                last.trailingTrivia.trimmingSuffix(while: \.isSpaceOrTab)
+              )
+            ]
+        }
```

Comment:
Not sure if you agree but I think this would read nicer as

```swift
} else if let lastAttribute = filteredAttributes.last {
  filteredAttributes[filteredAttributes.count - 1].trailingTrivia = lastAttribute.trailingTrivia.trimmingSuffix(while: \.isSpaceOrTab)
}
```